### DISCUSSION TITLE
failed: failed to set ssh key: The key already exists (KEY_ALREADY_EXISTS)

### DISCUSSION THREAD

**Discussion by Stringls** [syself/cluster-api-provider-hetzner#703] [Q&A]
Hi! It seems like BM server wasn't deprovisioned completely or at all and when I try to setup the cluster again the BM host is always in the preparing state. CAPH controller gives the below error
```
action \"preparing\" failed: failed to set ssh key: The key already exists (KEY_ALREADY_EXISTS)","errorVerbose":"The key already exists (KEY_ALREADY_EXISTS)\nfailed to set ssh key
```
How to reset the ssh key or deprovision BM server so I can setup it again?

**Comment by janiskemper** [comment]
The problem is here probably that the key exists in Robot with another name. Same SSH key, but a different name. Can you please check that? This is an issue that we have experienced multiple times as well when we named the key wrongly and this happens...

**Comment by Stringls** [comment]
Yeah, I've changed the Hetzner project and using another key. The name if different, not sure about the key data

**Reply by janiskemper** [reply]
Then I have the strong suggestion that you didn't change the data :smiley: Try with a different one!

**Reply by Stringls** [reply]
I see. If I found a right key for the BM server and cluster is created, and deprovision it successfully using `delete` then would I be able to use a different key with different name?

**Reply by janiskemper** [reply]
No, we have not implemented a deletion of SSH keys.. 
But that's a good question - I forgot to say something about it. 
SSH keys will be uploaded to Hetzner robot if they don't exist yet, but they will not be deleted afterwards. 

I guess that this shouldn't be the case because it is rather dangerous - if another cluster uses them. 

What do you think?

**Reply by janiskemper** [reply]
So to sum up: The unsuccessful delete has absolutely nothing to do with this problem with ssh keys AFAIK

**Reply by Stringls** [reply]
@janiskemper Am I able to delete a SSH key from Hetzner Robot and use another one for BM? I just cannot check as I do not have access.

I am not good at security questions, but I think so. For me it's just a problem that I cannot use another key and I cannot find the previous one :(

**Reply by janiskemper** [reply]
You can use another key, as long as it is actually another SSH KEY. Just renaming it will lead to this error. 
Using the same key as before with the same name will also work. 

If you can't delete the first key, then generate a new one and use that one. The controller will automatically upload it and as long as this key is not already uploaded (i.e. has the same fingerprint as an existing one), it will work.

**Reply by Stringls** [reply]
I am a bit confused. In Hcloud I have SSH key which name I pass in `HetznerCluster` for what do I need this key?
In `robot-ssh` I pass privatekey, publickey and a name. The name will be added to Hetzner Robot
I assume that robot-ssh is used for provisioning BM and login to it, while Hcloud SSH is used to login to Hcloud VM?

I just cannot figure out which name of SSH key I should use

**Reply by janiskemper** [reply]
In HCloud, you don't necessarily have to specify a key - only if you want. 
In Robot we need it to provision the server. As we have access to the full (public) key, we can also upload it with the controller in case the user didn't do it yet. That's what we do.
You can name the ssh key however you want! 

I don't really understand the problem you have with creating a new one and specifying it in the specs. Can you elaborate?

**Reply by Stringls** [reply]
All good right now, I've managed out how it works, thanks.

> In HCloud, you don't necessarily have to specify a key - only if you want.

In the [docs](https://github.com/syself/cluster-api-provider-hetzner/blob/main/docs/reference/hetzner-cluster.md) it says that we have to specify HCloud SSH key name.
```
sshKeys.hcloud.name | string |   | yes | Name of SSH key
```
So I am not sure what you mean by it's not necessary

**Reply by janiskemper** [reply]
The name of the ssh key you specified in HCloud is required here!

**Reply by janiskemper** [reply]
But I'm very happy that you made it work!

**Reply by Stringls** [reply]
Thank you for help! I'll close the discussion as completed
